File: bundles.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

const contentRootRel = "content"

// defaultContentBundleRel is the default content dir relative to the project/install root (see resolveProjectRoot).
const defaultContentBundleRel = "content/kometa"

// resolveProjectRoot picks the directory that contains Template/ and the default content bundle when the
// user runs the binary from another working directory (e.g. ~/ with the executable in the repo root).
// Order: cwd if it contains content/kometa/site.json, else executable directory if it does, else cwd.
func resolveProjectRoot() (string, error) {
	wd, errWd := os.Getwd()
	if errWd != nil {
		wd = ""
	}
	if wd != "" && contentBundleMarkerExists(wd) {
		return filepath.Clean(wd), nil
	}
	exe, err := os.Executable()
	if err != nil {
		if wd != "" {
			return filepath.Clean(wd), nil
		}
		return "", err
	}
	exe, err = filepath.EvalSymlinks(exe)
	if err != nil {
		return "", err
	}
	exeDir := filepath.Dir(exe)
	if contentBundleMarkerExists(exeDir) {
		return filepath.Clean(exeDir), nil
	}
	if wd != "" {
		return filepath.Clean(wd), nil
	}
	return filepath.Clean(exeDir), nil
}

func contentBundleMarkerExists(root string) bool {
	st, err := os.Stat(filepath.Join(root, filepath.FromSlash(defaultContentBundleRel), "site.json"))
	return err == nil && !st.IsDir()
}

// discoverContentBundles returns project-relative paths (e.g. content/kometa) for each
// subdirectory of content/ that contains a parseable site.json with site_id and output_folder.
func discoverContentBundles(projectRoot string) ([]string, error) {
	contentDir := filepath.Join(projectRoot, contentRootRel)
	entries, err := os.ReadDir(contentDir)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, fmt.Errorf("cannot read %q: %w", contentDir, err)
	}

	var bundles []string
	for _, ent := range entries {
		if !ent.IsDir() {
			continue
		}
		rel := filepath.ToSlash(filepath.Join(contentRootRel, ent.Name()))
		siteJSON := filepath.Join(projectRoot, filepath.FromSlash(rel), "site.json")
		if isValidSiteJSON(siteJSON) {
			bundles = append(bundles, rel)
		}
	}
	sort.Strings(bundles)
	return bundles, nil
}

func isValidSiteJSON(path string) bool {
	st, err := os.Stat(path)
	if err != nil || st.IsDir() {
		return false
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return false
	}
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		return false
	}
	_, hasID := raw["site_id"]
	_, hasOut := raw["output_folder"]
	return hasID && hasOut
}

func resolveSiteDir(projectRoot, siteInput string) string {
	siteInput = strings.TrimSpace(siteInput)
	if siteInput == "" {
		return filepath.Join(projectRoot, filepath.FromSlash(defaultContentBundleRel))
	}
	if filepath.IsAbs(siteInput) {
		return siteInput
	}
	return filepath.Join(projectRoot, filepath.FromSlash(siteInput))
}

File: crates/core/src/bundles.rs
use crate::error::{CoreError, CoreResult};
use std::fs;
use std::path::{Path, PathBuf};

const CONTENT_ROOT_REL: &str = "content";
const DEFAULT_CONTENT_BUNDLE_REL: &str = "content/kometa";

pub fn resolve_project_root() -> CoreResult<PathBuf> {
    let wd = std::env::current_dir().unwrap_or_else(|_| PathBuf::from("."));
    if content_bundle_marker_exists(&wd) {
        return Ok(wd);
    }
    if let Ok(exe) = std::env::current_exe() {
        let exe_dir = exe.parent().unwrap_or(Path::new(".")).to_path_buf();
        if content_bundle_marker_exists(&exe_dir) {
            return Ok(exe_dir);
        }
    }
    Ok(wd)
}

fn content_bundle_marker_exists(root: &Path) -> bool {
    root.join(DEFAULT_CONTENT_BUNDLE_REL)
        .join("site.json")
        .is_file()
}

pub fn discover_content_bundles(project_root: &Path) -> CoreResult<Vec<String>> {
    let content_dir = project_root.join(CONTENT_ROOT_REL);
    let entries = match fs::read_dir(&content_dir) {
        Ok(e) => e,
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => return Ok(vec![]),
        Err(e) => {
            return Err(CoreError::msg(format!(
                "cannot read {:?}: {e}",
                content_dir
            )))
        }
    };

    let mut bundles = Vec::new();
    for ent in entries.flatten() {
        if !ent.file_type().map(|t| t.is_dir()).unwrap_or(false) {
            continue;
        }
        let dir_name = ent.file_name();
        let dir_name = dir_name.to_string_lossy();
        if dir_name.starts_with('_') {
            continue;
        }
        let rel = format!("{}/{}", CONTENT_ROOT_REL, dir_name);
        let site_json = project_root.join(&rel).join("site.json");
        if is_valid_site_json(&site_json) {
            bundles.push(rel.replace('\\', "/"));
        }
    }
    bundles.sort();
    Ok(bundles)
}

fn is_valid_site_json(path: &Path) -> bool {
    let data = match fs::read_to_string(path) {
        Ok(d) => d,
        Err(_) => return false,
    };
    let raw: serde_json::Map<String, serde_json::Value> = match serde_json::from_str(&data) {
        Ok(v) => v,
        Err(_) => return false,
    };
    raw.contains_key("site_id") && raw.contains_key("output_folder")
}

pub fn resolve_site_dir(project_root: &Path, site_input: &str) -> PathBuf {
    let site_input = site_input.trim();
    if site_input.is_empty() {
        return project_root.join(DEFAULT_CONTENT_BUNDLE_REL);
    }
    let p = Path::new(site_input);
    if p.is_absolute() {
        p.to_path_buf()
    } else {
        project_root.join(site_input.replace('/', std::path::MAIN_SEPARATOR_STR))
    }
}
